api/v1beta1/metal3remediation_webhook.go

```diff
@@ -17,80 +17,93 @@ limitations under the License.
+func (webhook *Metal3Remediation) SetupWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
+		For(webhook).
+		WithDefaulter(webhook, admission.DefaulterRemoveUnknownOrOmitableFields).
+		WithValidator(webhook).
 		Complete()
 }
 
-// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediation,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations,versions=v1beta1,name=validation.metal3remediation.infrastructure.cluster.x-k8s.io,matchPolicy=Equivalent,sideEffects=None,admissionReviewVersions=v1;v1beta1
-// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediation,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations,versions=v1beta1,name=default.metal3remediation.infrastructure.cluster.x-k8s.io,matchPolicy=Equivalent,sideEffects=None,admissionReviewVersions=v1;v1beta1
-
-var _ webhook.Defaulter = &Metal3Remediation{}
-var _ webhook.Validator = &Metal3Remediation{}
+var _ webhook.CustomDefaulter = &Metal3Remediation{}
+var _ webhook.CustomValidator = &Metal3Remediation{}
 
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (r *Metal3Remediation) Default() {
+// Deprecated: This method is going to be removed in a next release.
+func (webhook *Metal3Remediation) Default(_ context.Context, _ runtime.Object) error {
+	return nil
 }
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (r *Metal3Remediation) ValidateCreate() (admission.Warnings, error) {
-	return nil, r.validate()
+// Deprecated: This method is going to be removed in a next release.
+func (webhook *Metal3Remediation) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
+	c, ok := obj.(*Metal3Remediation)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a Metal3Remediation but got a %T", obj))
+	}
+	return nil, webhook.validate(c)
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (r *Metal3Remediation) ValidateUpdate(_ runtime.Object) (admission.Warnings, error) {
-	return nil, r.validate()
+// Deprecated: This method is going to be removed in a next release.
+func (webhook *Metal3Remediation) ValidateUpdate(_ context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	newM3R, ok := newObj.(*Metal3Remediation)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a Metal3Remediation but got a %T", newObj))
+	}
+
+	return nil, webhook.validate(newM3R)
 }
 
 // ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
-func (r *Metal3Remediation) ValidateDelete() (admission.Warnings, error) {
-	metal3remediationlog.Info("validate delete", "name", r.Name)
```

Comment:
do we need to remove this logline as well?

Reply:
It looked odd since we aren't logging ValidateDelete() anywhere else, except here. That's why I removed this log.